diff -Naur trac-0.10.3-vanilla/trac/attachment.py trac-0.10.3/trac/attachment.py
--- trac-0.10.3-vanilla/trac/attachment.py	2006-12-12 19:44:08.000000000 +0100
+++ trac-0.10.3/trac/attachment.py	2007-03-06 21:49:08.000000000 +0100
@@ -524,7 +524,10 @@
                                  'author': get_reporter_id(req)}
 
     def _render_view(self, req, attachment):
-        perm_map = {'ticket': 'TICKET_VIEW', 'wiki': 'WIKI_VIEW'}
+        if re.search(r'Public', attachment.description):
+            perm_map = {'ticket': 'TICKET_VIEW', 'wiki': 'WIKI_PUBLIC'}
+        else:
+            perm_map = {'ticket': 'TICKET_VIEW', 'wiki': 'WIKI_VIEW'}
         req.perm.assert_permission(perm_map[attachment.parent_type])
 
         req.check_modified(attachment.time)
diff -Naur trac-0.10.3-vanilla/trac/wiki/web_ui.py trac-0.10.3/trac/wiki/web_ui.py
--- trac-0.10.3-vanilla/trac/wiki/web_ui.py	2006-12-12 19:44:07.000000000 +0100
+++ trac-0.10.3/trac/wiki/web_ui.py	2007-03-06 16:20:48.000000000 +0100
@@ -73,7 +73,7 @@
     # IPermissionRequestor methods
 
     def get_permission_actions(self):
-        actions = ['WIKI_CREATE', 'WIKI_DELETE', 'WIKI_MODIFY', 'WIKI_VIEW']
+        actions = ['WIKI_CREATE', 'WIKI_DELETE', 'WIKI_MODIFY', 'WIKI_VIEW', 'WIKI_PUBLIC']
         return actions + [('WIKI_ADMIN', actions)]
 
     # IRequestHandler methods
@@ -127,7 +127,10 @@
         elif action == 'history':
             self._render_history(req, db, page)
         else:
-            req.perm.assert_permission('WIKI_VIEW')            
+            if re.search(r'Public', page.name):
+                req.perm.assert_permission('WIKI_PUBLIC')
+            else:
+                req.perm.assert_permission('WIKI_VIEW')
             format = req.args.get('format')
             if format:
                 Mimeview(self.env).send_converted(req, 'text/x-trac-wiki',
@@ -263,7 +266,10 @@
                                'only_version': num_versions == 1}
 
     def _render_diff(self, req, db, page):
-        req.perm.assert_permission('WIKI_VIEW')
+        if re.search(r'Public', page.name):
+            req.perm.assert_permission('WIKI_PUBLIC')
+        else:
+            req.perm.assert_permission('WIKI_VIEW')
 
         if not page.exists:
             raise TracError("Version %s of page %s does not exist" %
