Index: trac/env.py
===================================================================
--- trac.ori/env.py (revision 314)
+++ trac/env.py (revision 315)
@@ -104,0 +104,33 @@
+    tr_fixed = Option('translation', 'fixed', u'opgelost', '')
+    tr_invalid = Option('translation', 'invalid', u'ongeldig', '')
+    tr_wontfix = Option('translation', 'wontfix', u'onoplosbaar', '')
+    tr_duplicate = Option('translation', 'duplicate', u'duplicaat', '')
+    tr_worksforme = Option('translation', 'worksforme', u'nietreproduceerbaar', '')
+    tr_new = Option('translation', 'new', u'nieuw', '')
+    tr_closed = Option('translation', 'closed', u'gesloten', '')
+    tr_assigned = Option('translation', 'assigned', u'toegekend', '')
+    tr_reopened = Option('translation', 'reopened', u'heropend', '')
+    tr_reassign = Option('translation', 'reassign', u'opnieuwtoegekend', '')
+    tr_blocker = Option('translation', 'blocker', u'blokkering', '')
+    tr_critical = Option('translation', 'critical', u'kritisch', '')
+    tr_major = Option('translation', 'major', u'majeur', '')
+    tr_minor = Option('translation', 'minor', u'mineur', '')
+    tr_trivial = Option('translation', 'trivial', u'triviaal', '')
+    tr_defect = Option('translation', 'defect', u'defect', '')
+    tr_enhancement = Option('translation', 'enhancement', u'verbetering', '')
+    tr_task = Option('translation', 'task', u'taak', '')
+    tr_component = Option('translation', 'component', u'onderdeel', '')
+    tr_keywords = Option('translation', 'keywords', u'sleutelwoorden', '')
+    tr_priority = Option('translation', 'priority', u'prioriteit', '')
+    tr_milestone = Option('translation', 'milestone', u'mijlpaal', '')
+    tr_summary = Option('translation', 'summary', u'samenvatting', '')
+    tr_resolution = Option('translation', 'resolution', u'oplossing', '')
+    tr_report = Option('translation', 'report', u'rapport', '')
+    tr_title = Option('translation', 'title', u'titel', '')
+    tr_owner = Option('translation', 'owner', u'eigenaar', '')
+    tr_reporter = Option('translation', 'reporter', u'rapporteur', '')
+    tr_created = Option('translation', 'created', u'aangemaakt', '')
+    tr_modified = Option('translation', 'modified', u'aangepast', '')
+    tr_change = Option('translation', 'change', u'wijziging', '')
+    tr_changeset = Option('translation', 'changeset', u'versie', '')
+
@@ -127,0 +160,1 @@
+            self.init_translations()
@@ -249,0 +283,8 @@
+
+    def init_translations(self):
+        self.translations = {}
+        for (k,v) in self.config.options('translation'):
+            self.translations[k] = v
+            ck = k.capitalize()
+            cv = v.capitalize()
+            self.translations[ck] = cv
Index: trac/core.py
===================================================================
--- trac.ori/core.py (revision 314)
+++ trac/core.py (revision 315)
@@ -177,1 +177,1 @@
-                raise TracError, 'Component "%s" not registered' % cls.__name__
+                raise TracError, 'Component "%s" niet geregistreerd' % cls.__name__
@@ -181,1 +181,1 @@
-                raise TracError, 'Unable to instantiate component %r (%s)' \
+                raise TracError, 'Kan component niet laden %r (%s)' \
Index: trac/attachment.py
===================================================================
--- trac.ori/attachment.py (revision 314)
+++ trac/attachment.py (revision 315)
@@ -103,2 +103,2 @@
-            raise TracError('Attachment %s does not exist.' % (self.title),
-                            'Invalid Attachment')
+            raise TracError('Bijlage %s bestaat niet.' % (self.title),
+                            'Ongeldige bijlage')
@@ -133,1 +133,1 @@
-        assert self.filename, 'Cannot delete non-existent attachment'
+        assert self.filename, 'Kan niet bestaande bijlage niet verwijderen'
@@ -152,1 +152,1 @@
-                raise TracError, 'Could not delete attachment'
+                raise TracError, 'Kan bijlage niet verwijderen'
@@ -253,1 +253,1 @@
-            raise TracError('Attachment %s not found' % self.filename)
+            raise TracError('Bijlage %s niet gevonden' % self.filename)
@@ -339,1 +339,1 @@
-            raise HTTPBadRequest('Bad request')
+            raise HTTPBadRequest('Ongeldig verzoek')
@@ -341,1 +341,1 @@
-            raise HTTPBadRequest('Unknown attachment type')
+            raise HTTPBadRequest('Bestandstype bijlage onbekend')
@@ -354,1 +354,1 @@
-                raise HTTPBadRequest('Bad request')
+                raise HTTPBadRequest('Ongeldig verzoek')
@@ -379,1 +379,1 @@
-            parent_text = 'Ticket #' + parent_id
+            parent_text = 'Melding #' + parent_id
@@ -430,1 +430,1 @@
-                title += Markup(' by %s', author)
+                title += Markup(' door %s', author)
@@ -446,1 +446,1 @@
-            raise TracError('No file uploaded')
+            raise TracError('Geen bestand ontvangen')
@@ -452,1 +452,1 @@
-            raise TracError("Can't upload empty file")
+            raise TracError("Kan leeg bestand niet versturen")
@@ -457,2 +457,2 @@
-            raise TracError('Maximum attachment size: %d bytes' % max_size,
-                            'Upload failed')
+            raise TracError('Maximale bestandsgrootte van bijlagen: %d bytes' % max_size,
+                            'Upload mislukt')
@@ -467,1 +467,1 @@
-            raise TracError('No file uploaded')
+            raise TracError('Geen bestand verstuurd')
@@ -477,1 +477,1 @@
-                    raise InvalidAttachment('Attachment field %s is invalid: %s'
+                    raise InvalidAttachment('Veld %s van bijlage is ongeldig: %s'
@@ -480,1 +480,1 @@
-                    raise InvalidAttachment('Invalid attachment: %s' % message)
+                    raise InvalidAttachment('Ongeldige bijlage: %s' % message)
@@ -515,1 +515,1 @@
-        req.hdf['title'] = '%s (delete)' % attachment.title
+        req.hdf['title'] = '%s (verwijder)' % attachment.title
@@ -575,1 +575,1 @@
-                add_link(req, 'alternate', plaintext_href, 'Plain Text',
+                add_link(req, 'alternate', plaintext_href, 'Standaard tekst',
@@ -580,1 +580,1 @@
-            add_link(req, 'alternate', raw_href, 'Original Format', mime_type)
+            add_link(req, 'alternate', raw_href, 'Origineel formaat', mime_type)
@@ -621,1 +621,1 @@
-                          title='Attachment %s' % attachment.title)
+                          title='Bijlage %s' % attachment.title)
Index: trac/siteconfig.py
===================================================================
--- trac.ori/siteconfig.py (revision 314)
+++ trac/siteconfig.py (revision 315)
@@ -3,1 +3,1 @@
-# This file was autogenerated when installing Trac 0.10.3.
+# This file was autogenerated when installing Trac 0.10.
Index: trac/ticket/api.py
===================================================================
--- trac.ori/ticket/api.py (revision 314)
+++ trac/ticket/api.py (revision 315)
Index: trac/ticket/web_ui.py
===================================================================
--- trac.ori/ticket/web_ui.py (revision 314)
+++ trac/ticket/web_ui.py (revision 315)
@@ -32,1 +32,1 @@
-from trac.util.text import CRLF
+from trac.util.text import CRLF, translate
@@ -110,1 +110,1 @@
-               html.A('New Ticket', href=req.href.newticket(), accesskey=7))
+               html.A('Nieuwe melding', href=req.href.newticket(), accesskey=7))
@@ -137,1 +137,1 @@
-        req.hdf['title'] = 'New Ticket'
+        req.hdf['title'] = 'Nieuwe melding'
@@ -159,1 +159,1 @@
-                field['label'] = 'Assign to'
+                field['label'] = 'Toekennen aan'
@@ -170,0 +170,2 @@
+
+            field['label'] = translate(self.env, field['label'])
@@ -235,1 +237,1 @@
-        yield ('csv', 'Comma-delimited Text', 'csv',
+        yield ('csv', 'Komma-gescheiden tekst', 'csv',
@@ -237,1 +239,1 @@
-        yield ('tab', 'Tab-delimited Text', 'tsv',
+        yield ('tab', 'TAB-gescheiden tekst', 'tsv',
@@ -318,1 +320,1 @@
-                             'Ticket #%s' % tickets[0])
+                             'Melding #%s' % tickets[0])
@@ -320,1 +322,1 @@
-                             'Ticket #%s' % tickets[idx - 1])
+                             'Melding #%s' % tickets[idx - 1])
@@ -323,1 +325,1 @@
-                             'Ticket #%s' % tickets[idx + 1])
+                             'Melding #%s' % tickets[idx + 1])
@@ -325,1 +327,1 @@
-                             'Ticket #%s' % tickets[-1])
+                             'Melding #%s' % tickets[-1])
@@ -342,1 +344,1 @@
-            yield ('ticket', 'Ticket changes')
+            yield ('ticket', 'Aanpassingen van melding')
@@ -344,1 +346,1 @@
-                yield ('ticket_details', 'Ticket details', False)
+                yield ('ticket_details', 'Details van melding', False)
@@ -349,4 +351,4 @@
-        status_map = {'new': ('newticket', 'created'),
-                      'reopened': ('newticket', 'reopened'),
-                      'closed': ('closedticket', 'closed'),
-                      'edit': ('editedticket', 'updated')}
+        status_map = {'new': ('newticket', 'gemaakt'),
+                      'reopened': ('newticket', 'heropend'),
+                      'closed': ('closedticket', 'gesloten'),
+                      'edit': ('editedticket', 'bijgewerkt')}
@@ -363,1 +365,1 @@
-                                          fields.keys()]) + ' changed<br />'
+                                          fields.keys()]) + ' gewijzigd<br />'
@@ -377,2 +379,2 @@
-                title = 'Ticket #%s (%s %s): %s' % \
-                        (id, type.lower(), verb, summary)
+                title = 'Melding #%s (%s %s): %s' % \
+                        (id, translate(self.env, type).lower(), verb, summary)
@@ -380,2 +382,2 @@
-                title = Markup('Ticket <em title="%s">#%s</em> (%s) %s by %s',
-                               summary, id, type, verb, author)
+                title = Markup('Melding <em title="%s">#%s</em> (%s) %s par %s',
+                               summary, id, translate(self.env, type), verb, author)
@@ -609,1 +611,1 @@
-                    ['Replying to [%s %s]:' % (link, author)] +
+                    ['Reagerend op [%s %s]:' % (link, author)] +
Index: trac/ticket/report.py
===================================================================
--- trac.ori/ticket/report.py (revision 314)
+++ trac/ticket/report.py (revision 315)
@@ -49,1 +50,1 @@
-               html.A('View Tickets', href=req.href.report()))
+               html.A('Meldingen weergeven', href=req.href.report()))
@@ -96,1 +97,1 @@
-            add_link(req, 'up', req.href.report(), 'Available Reports')
+            add_link(req, 'up', req.href.report(), 'Beschikbare rapporten')
Index: trac/ticket/model.py
===================================================================
--- trac.ori/ticket/model.py (revision 314)
+++ trac/ticket/model.py (revision 315)
@@ -86,2 +86,2 @@
-            raise TracError('Ticket %d does not exist.' % tkt_id,
-                            'Invalid Ticket Number')
+            raise TracError('Melding %d bestaat niet.' % tkt_id,
+                            'Ongeldig meldingsnummer')
@@ -134,1 +134,1 @@
-        assert not self.exists, 'Cannot insert an existing ticket'
+        assert not self.exists, 'Kan een reeds bestaande melding niet opnieuw invoeren'
@@ -190,1 +190,1 @@
-        assert self.exists, 'Cannot update a new ticket'
+        assert self.exists, 'Kan een nieuwe melding niet bijwerken'
@@ -334,1 +334,1 @@
-                raise TracError, '%s %s does not exist.' % (self.type, name)
+                raise TracError, '%s %s bestaat niet.' % (self.type, name)
@@ -344,1 +344,1 @@
-        assert self.exists, 'Cannot deleting non-existent %s' % self.type
+        assert self.exists, 'Kan niet bestaande melding %s niet verwijderen' % self.type
@@ -362,2 +362,2 @@
-        assert not self.exists, 'Cannot insert existing %s' % self.type
-        assert self.name, 'Cannot create %s with no name' % self.type
+        assert not self.exists, 'Kan reeds bestaande %s niet invoegen' % self.type
+        assert self.name, 'Kan %s niet maken zonder naam' % self.type
@@ -387,2 +387,2 @@
-        assert self.exists, 'Cannot update non-existent %s' % self.type
-        assert self.name, 'Cannot update %s with no name' % self.type
+        assert self.exists, 'Kan niet bestaande melding %s niet bijwerken' % self.type
+        assert self.name, 'Kan %s zonder naam niet bijwerken' % self.type
@@ -460,1 +460,1 @@
-                raise TracError, 'Component %s does not exist.' % name
+                raise TracError, 'Onderdeel %s bestaat niet.' % name
@@ -472,1 +472,1 @@
-        assert self.exists, 'Cannot deleting non-existent component'
+        assert self.exists, 'Kan een niet bestaand onderdeel niet verwijderen'
@@ -489,2 +489,2 @@
-        assert not self.exists, 'Cannot insert existing component'
-        assert self.name, 'Cannot create component with no name'
+        assert not self.exists, 'Kan reeds bestaand onderdeel niet invoegen'
+        assert self.name, 'Kan onderdeel zonder naam niet maken'
@@ -508,2 +508,2 @@
-        assert self.exists, 'Cannot update non-existent component'
-        assert self.name, 'Cannot update component with no name'
+        assert self.exists, 'Kan niet bestaand onderdeel niet bijwerken'
+        assert self.name, 'Kan onderdeel zonder naam niet bijwerken'
@@ -597,1 +597,1 @@
-            ticket.save_changes(author, 'Milestone %s deleted' % self.name,
+            ticket.save_changes(author, 'Mijlpaal %s verwijderd' % self.name,
@@ -604,1 +604,1 @@
-        assert self.name, 'Cannot create milestone with no name'
+        assert self.name, 'Kan mijlpaal zonder naam niet maken'
@@ -622,1 +622,1 @@
-        assert self.name, 'Cannot update milestone with no name'
+        assert self.name, 'Kan mijlpaal zonder naam niet bijwerken'
@@ -693,1 +693,1 @@
-        assert self.exists, 'Cannot deleting non-existent version'
+        assert self.exists, 'Kan niet bestaande versie niet verwijderen'
@@ -711,1 +711,1 @@
-        assert self.name, 'Cannot create version with no name'
+        assert self.name, 'Kan versie zonder naam niet maken'
@@ -729,2 +729,2 @@
-        assert self.exists, 'Cannot update non-existent version'
-        assert self.name, 'Cannot update version with no name'
+        assert self.exists, 'Kan niet bestaande versie niet bijwerken'
+        assert self.name, 'Kan versie zonder naam niet bijwerken'
Index: trac/ticket/roadmap.py
===================================================================
--- trac.ori/ticket/roadmap.py (revision 314)
+++ trac/ticket/roadmap.py (revision 315)
@@ -90,0 +90,3 @@
+    safe_name = None
+    if milestone.exists:
+        safe_name = milestone.name.replace('/', '%2F')
@@ -91,1 +94,1 @@
-           'href': req.href.milestone(milestone.name)}
+           'href': req.href.milestone(safe_name)}
@@ -132,1 +135,1 @@
-               html.a('Roadmap', href=req.href.roadmap(), accesskey=3))
+               html.a('Planning', href=req.href.roadmap(), accesskey=3))
@@ -146,1 +149,1 @@
-        req.hdf['title'] = 'Roadmap'
+        req.hdf['title'] = 'Planning'
@@ -246,1 +249,1 @@
-                write_utctime('DTSTAMP', localtime(milestone['due']))
+                write_date('DTSTAMP', localtime(milestone['due']))
@@ -292,1 +295,1 @@
-        return 'roadmap'
+        return 'planning'
@@ -309,1 +312,1 @@
-            yield ('milestone', 'Milestones')
+            yield ('milestone', 'Mijlpalen')
@@ -320,1 +323,1 @@
-                title = Markup('Milestone <em>%s</em> completed', name)
+                title = Markup('Mijlpaal <em>%s</em> gecompleteerd', name)
@@ -438,1 +440,1 @@
-        req.hdf['title'] = 'Milestone %s' % milestone.name
+        req.hdf['title'] = 'Mijlpaal %s' % milestone.name
@@ -450,1 +452,1 @@
-            req.hdf['title'] = 'Milestone %s' % milestone.name
+            req.hdf['title'] = 'Mijlpaal %s' % milestone.name
@@ -457,1 +459,1 @@
-            req.hdf['title'] = 'New Milestone'
+            req.hdf['title'] = 'Nieuwe mijlpaal'
@@ -468,1 +470,1 @@
-        req.hdf['title'] = 'Milestone %s' % milestone.name
+        req.hdf['title'] = 'Mijlpaal %s' % milestone.name
Index: trac/ticket/query.py
===================================================================
--- trac.ori/ticket/query.py (revision 314)
+++ trac/ticket/query.py (revision 315)
@@ -380,1 +380,1 @@
-                   html.A('View Tickets', href=req.href.query()))
+                   html.A('Meldingen weergeven', href=req.href.query()))
@@ -488,4 +488,4 @@
-            {'name': "contains", 'value': "~"},
-            {'name': "doesn't contain", 'value': "!~"},
-            {'name': "begins with", 'value': "^"},
-            {'name': "ends with", 'value': "$"},
+            {'name': "bevat", 'value': "~"},
+            {'name': "bevat niet", 'value': "!~"},
+            {'name': "begint met", 'value': "^"},
+            {'name': "eindigt met", 'value': "$"},
@@ -493,1 +493,1 @@
-            {'name': "is not", 'value': "!"}
+            {'name': "is niet", 'value': "!"}
@@ -497,1 +497,1 @@
-            {'name': "is not", 'value': "!"}
+            {'name': "is niet", 'value': "!"}
@@ -502,1 +502,1 @@
-        req.hdf['title'] = 'Custom Query'
+        req.hdf['title'] = 'Aangepaste zoekopdracht'
Index: trac/__init__.py
===================================================================
--- trac.ori/__init__.py (revision 314)
+++ trac/__init__.py (revision 315)
@@ -14,1 +14,1 @@
-__version__ = '0.10.3'
+__version__ = '0.10'
Index: trac/versioncontrol/web_ui/changeset.py
===================================================================
--- trac.ori/versioncontrol/web_ui/changeset.py (revision 314)
+++ trac/versioncontrol/web_ui/changeset.py (revision 315)
@@ -264,1 +264,1 @@
-        add_link(req, 'alternate', '?format=zip&'+diff_params, 'Zip Archive',
+        add_link(req, 'alternate', '?format=zip&'+diff_params, 'Zip-archief',
Index: trac/versioncontrol/web_ui/browser.py
===================================================================
--- trac.ori/versioncontrol/web_ui/browser.py (revision 314)
+++ trac/versioncontrol/web_ui/browser.py (revision 315)
@@ -69,1 +69,1 @@
-               html.A('Browse Source', href=req.href.browser()))
+               html.A('Bronbestanden', href=req.href.browser()))
Index: trac/Settings.py
===================================================================
--- trac.ori/Settings.py (revision 314)
+++ trac/Settings.py (revision 315)
@@ -36,1 +36,1 @@
-               html.A('Settings', href=req.href.settings()))
+               html.A('Instellingen', href=req.href.settings()))
@@ -52,1 +52,1 @@
-        req.hdf['title'] = 'Settings'
+        req.hdf['title'] = 'Instellingen'
Index: trac/Search.py
===================================================================
--- trac.ori/Search.py (revision 314)
+++ trac/Search.py (revision 315)
@@ -77,1 +77,1 @@
-        raise TracError('Empty search attempt, this should really not happen.')
+        raise TracError('Leeg zoekresultaat; dit zou niet moeten gebeuren.')
@@ -135,1 +135,1 @@
-               html.A('Search', href=req.href.search(), accesskey=4))
+               html.A('Zoeken', href=req.href.search(), accesskey=4))
@@ -165,1 +165,1 @@
-        req.hdf['title'] = 'Search'
+        req.hdf['title'] = 'Zoeken'
@@ -176,3 +176,3 @@
-                raise TracError('Search query too short. '
-                                'Query must be at least %d characters long.' % \
-                                self.min_query_length, 'Search Error')
+                raise TracError('Zoekterm te kort. '
+                                'Een zoekopdracht moet uit minimaal %d karakters bestaan.' % \
+                                self.min_query_length, 'Zoekfout')
@@ -188,1 +188,1 @@
-            req.hdf['title'] = 'Search Results'
+            req.hdf['title'] = 'Zoekresultaten'
@@ -198,1 +198,1 @@
-                add_link(req, 'next', next_href, 'Next Page')
+                add_link(req, 'next', next_href, 'Volgende pagina')
@@ -203,1 +203,1 @@
-                add_link(req, 'prev', prev_href, 'Previous Page')
+                add_link(req, 'prev', prev_href, 'Vorige pagina')
@@ -225,1 +225,1 @@
-            description = 'Browse repository path ' + kwd
+            description = 'Blader door repository ' + kwd
Index: trac/Timeline.py
===================================================================
--- trac.ori/Timeline.py (revision 314)
+++ trac/Timeline.py (revision 315)
@@ -80,1 +80,1 @@
-               html.A('Timeline', href=req.href.timeline(), accesskey=2))
+               html.A('Tijdslijn', href=req.href.timeline(), accesskey=2))
@@ -154,1 +154,1 @@
-        req.hdf['title'] = 'Timeline'
+        req.hdf['title'] = 'Tijdslijn'
Index: trac/About.py
===================================================================
--- trac.ori/About.py (revision 314)
+++ trac/About.py (revision 315)
@@ -41,1 +41,1 @@
-               html.a('About Trac', href=req.href.about()))
+               html.a('Over Trac', href=req.href.about()))
@@ -59,1 +59,1 @@
-        req.hdf['title'] = 'About Trac'
+        req.hdf['title'] = 'Over Trac'
Index: trac/wiki/web_ui.py
===================================================================
--- trac.ori/wiki/web_ui.py (revision 314)
+++ trac/wiki/web_ui.py (revision 315)
@@ -70,1 +70,1 @@
-               html.A('Help/Guide', href=req.href.wiki('TracGuide'),
+               html.A('Hulp', href=req.href.wiki('hulp'),
@@ -145,1 +145,1 @@
-            yield ('wiki', 'Wiki changes')
+            yield ('wiki', 'Wijzigingen in Wiki')
@@ -158,1 +158,1 @@
-                title = Markup('<em>%s</em> edited by %s',
+                title = Markup('<em>%s</em> bewerkt door %s',
Index: trac/test.py
===================================================================
--- trac.ori/test.py (revision 314)
+++ trac/test.py (revision 315)
Index: trac/util/__init__.py
===================================================================
--- trac.ori/util/__init__.py (revision 314)
+++ trac/util/__init__.py (revision 315)
@@ -33,1 +33,1 @@
-                           wrap, pretty_size
+                           wrap, pretty_size, translate
Index: trac/util/datefmt.py
===================================================================
--- trac.ori/util/datefmt.py (revision 314)
+++ trac/util/datefmt.py (revision 315)
@@ -33,6 +33,6 @@
-    units = ((3600 * 24 * 365, 'year',   'years'),
-             (3600 * 24 * 30,  'month',  'months'),
-             (3600 * 24 * 7,   'week',   'weeks'),
-             (3600 * 24,       'day',    'days'),
-             (3600,            'hour',   'hours'),
-             (60,              'minute', 'minutes'))
+    units = ((3600 * 24 * 365, 'jaar',   'jaren'),
+             (3600 * 24 * 30,  'maand',  'maanden'),
+             (3600 * 24 * 7,   'week',   'weken'),
+             (3600 * 24,       'dag',    'dagen'),
+             (3600,            'uur',   'uren'),
+             (60,              'minuut', 'minuten'))
@@ -43,1 +43,1 @@
-        return '%i second%s' % (age_s, age_s != 1 and 's' or '')
+        return '%i seconde%s' % (age_s, age_s != 1 and 'n' or '')
Index: trac/util/html.py
===================================================================
--- trac.ori/util/html.py (revision 314)
+++ trac/util/html.py (revision 315)
Index: trac/util/text.py
===================================================================
--- trac.ori/util/text.py (revision 314)
+++ trac/util/text.py (revision 315)
@@ -149,0 +149,9 @@
+
+# -- Translation
+
+def translate(env, name, capitalize=False):
+    if capitalize:
+        return env.translations.get(name.lower().capitalize(), name.capitalize())
+    else:
+        return env.translations.get(name.lower(), name)
+
Index: trac/web/chrome.py
===================================================================
--- trac.ori/web/chrome.py (revision 314)
+++ trac/web/chrome.py (revision 315)
@@ -280,1 +280,1 @@
-        add_link(req, 'help', req.href.wiki('TracGuide'))
+        add_link(req, 'help', req.href.wiki('hulp'))
Index: trac/web/main.py
===================================================================
--- trac.ori/web/main.py (revision 314)
+++ trac/web/main.py (revision 315)
@@ -33,1 +33,1 @@
-from trac.util.text import to_unicode
+from trac.util.text import to_unicode, translate
