Index: trac/env.py
===================================================================
--- trac/env.py	(revision 7872)
+++ trac/env.py	(working copy)
@@ -594,9 +594,10 @@
             env.log.error("Exception caught while checking for upgrade: %s",
                           exception_to_unicode(e, traceback=True))
         if needs_upgrade:
-            raise TracError(_('The Trac Environment needs to be upgraded.\n\n'
-                              'Run "trac-admin %(path)s upgrade"',
-                              path=env_path))
+            try:
+                env.upgrade(backup=True)
+            except TracError, e:
+                env.upgrade()
 
     return env
 
