Index: setup.py
===================================================================
--- setup.py	(revision 1962)
+++ setup.py	(working copy)
@@ -211,7 +211,7 @@
       url=URL,
       packages=['trac', 'trac.mimeview', 'trac.scripts', 'trac.ticket',
                 'trac.upgrades', 'trac.versioncontrol', 'trac.web',
-                'trac.wiki'],
+                'trac.wiki', 'trac.admin'],
       data_files=[(_p('share/trac/templates'), glob('templates/*')),
                   (_p('share/trac/htdocs'), glob(_p('htdocs/*.*')) + [_p('htdocs/README')]),
                   (_p('share/trac/htdocs/css'), glob(_p('htdocs/css/*'))),
Index: trac/scripts/admin.py
===================================================================
--- trac/scripts/admin.py	(revision 1962)
+++ trac/scripts/admin.py	(working copy)
@@ -39,6 +39,9 @@
 from trac.Milestone import Milestone
 from trac.perm import PermissionSystem
 from trac.ticket.model import *
+from trac.admin.console import IAdminConsoleProvider
+from trac.core import ComponentManager, ExtensionPoint
+from trac.loader import load_components
 
 try:
     sum
@@ -50,7 +53,24 @@
             tot += item
         return tot
 
+class AdminCommands(trac.core.Component):
+    """
+    Component end-point for IAdminConsoleProvider extensions
+    """
+    admin_providers = ExtensionPoint(IAdminConsoleProvider)
 
+    def import_providers(self, env, tracadm):
+        load_components(env)
+        for provider in self.admin_providers:
+            for command in provider.get_console_commands(tracadm):
+                name, help, callback, completer = command
+                if hasattr(tracadm.__class__, 'do_' + name):
+                    raise Exception("Can't use IAdminConsoleProvider %s, console command '%s' already exists" % (provider.__class__.__name__, name))
+                setattr(tracadm.__class__, 'do_' + name, callback)
+                setattr(tracadm.__class__, '_help_' + name, help)
+                if completer:
+                    setattr(tracadm.__class__, 'complete_' + name, completer)
+
 class TracAdmin(cmd.Cmd):
     intro = ''
     license = trac.__license_long__
@@ -64,6 +84,7 @@
     _date_format = '%Y-%m-%d'
     _datetime_format = '%Y-%m-%d %H:%M:%S'
     _date_format_hint = 'YYYY-MM-DD'
+    _admin_commands = None
 
     def __init__(self, envdir=None):
         cmd.Cmd.__init__(self)
@@ -72,6 +93,7 @@
             self.env_set(os.path.abspath(envdir))
 
     def docmd(self, cmd='help'):
+        self.env_check()
         self.onecmd(cmd)
 
     def emptyline(self):
@@ -84,6 +106,7 @@
               '%(copy)s\n\n'                                    \
               "Type:  '?' or 'help' for help on commands.\n" %  \
               {'ver':trac.__version__,'copy':__copyright__}
+        self.env_check()
         self.cmdloop()
 
     ##
@@ -99,6 +122,12 @@
     def env_check(self):
         try:
             self.__env = Environment(self.envname)
+            try:
+                self._admin_commands = AdminCommands(self.__env)
+                self._admin_commands.import_providers(self.__env, self)
+            except Exception, e:
+                print "Failed to initialise extension points.", e
+                raise
         except:
             return 0
         return 1
@@ -117,6 +146,8 @@
         try:
             if not self.__env:
                 self.__env = Environment(self.envname)
+                self._admin_commands = AdminCommands(self.__env)
+                self._admin_commands.import_providers(self.__env, self)
             return self.__env
         except Exception, e:
             print 'Failed to open environment.', e
@@ -209,6 +240,8 @@
                                xrange(0, (1 + len(sep)) * cnum + sum(colw))])
         print
 
+
+
     def print_doc(self, doc, decor=False):
         if not doc: return
         self.print_listing(['Command', 'Description'], doc, '  --', decor) 
@@ -299,15 +332,13 @@
             except AttributeError:
                 print "No documentation found for '%s'" % arg[0]
         else:
-            docs = (self._help_about + self._help_help +
-                    self._help_initenv + self._help_hotcopy +
-                    self._help_resync + self._help_upgrade +
-                    self._help_wiki +
-#                    self._help_config + self._help_wiki +
-                    self._help_permission + self._help_component +
-                    self._help_ticket_type + self._help_priority +
-                    self._help_severity +  self._help_version +
-                    self._help_milestone)
+            # Extract documentation from all _help_* members
+            docs = []
+            doc_strings = [doc for doc in dir(self) if doc.startswith('_help_') and doc[6:] not in ('EOF', 'exit', 'quit')]
+            for doc in doc_strings:
+                docs.extend(getattr(self, doc))
+            docs.sort(lambda a, b: cmp(a[0], b[0]))
+
             print 'trac-admin - The Trac Administration Console %s' % trac.__version__
             if not self.interactive:
                 print
Index: trac/admin/__init__.py
===================================================================
--- trac/admin/__init__.py	(revision 0)
+++ trac/admin/__init__.py	(revision 0)
@@ -0,0 +1 @@
+from trac.admin.console import *
Index: trac/admin/console.py
===================================================================
--- trac/admin/console.py	(revision 0)
+++ trac/admin/console.py	(revision 0)
@@ -0,0 +1,41 @@
+# -*- coding: iso8859-1 -*-
+#
+# Copyright (C) 2003, 2004, 2005 Edgewall Software
+# Copyright (C) 2003, 2004, 2005 Alec Thomas <alec@swapoff.org>
+#
+# Trac is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Trac is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+#
+# Author: Alec Thomas <alec@swapoff.org>
+
+from trac.core import Interface
+import sys
+
+__all__ = [ 'IAdminConsoleProvider' ]
+
+class IAdminConsoleProvider(Interface):
+    """
+    Extension point interface for components to provide an administrative
+    interface from within trac-admin.
+    """
+
+    def get_console_commands(tracadm):
+        """
+        Return an iterable of (name, help, callable, completer) tuples.
+
+        tracadm is a TracAdmin instance.
+
+        completer can be null. help is in the same format that trac-admin
+        uses.
+        """
