Index: trac/versioncontrol/cache.py
===================================================================
--- trac/versioncontrol/cache.py	(revision 11483)
+++ trac/versioncontrol/cache.py	(working copy)
@@ -40,6 +40,28 @@
 CACHE_METADATA_KEYS = (CACHE_REPOSITORY_DIR, CACHE_YOUNGEST_REV)
 
 
+class ICacheChangesetListener(Interface):                                  
+    """Cached changeset operations"""                                      
+                                                                           
+    def edit_changeset(cset):                                              
+        """Called when changeset is about to be cached.                    
+           Returns altered data to cache or None if unchanged. cset usually
+           contains cset.date, cset.author, cset.message and cset.rev      
+        """                                                                
+                                                                           
+class CacheManager(Component):                                             
+    """Provide interface to plug-in into cache operations"""               
+                                                                           
+    observers = ExtensionPoint(ICacheChangesetListener)                    
+                                                                           
+    def check_changeset(self, cset):                                       
+        for observer in self.observers:                                    
+            res = observer.edit_changeset(cset)                            
+            if res != None:                                                
+                cset = res                                                 
+        return cset                                                        
+
+
 class CachedRepository(Repository):
 
     has_linear_changesets = False
@@ -53,6 +75,7 @@
                              + CachedRepository.__name__ + '.metadata:'
                              + str(self.repos.id))
         Repository.__init__(self, repos.name, repos.params, log)
+        self.cache_mgr = CacheManager(env)
 
     def close(self):
         self.repos.close()
@@ -87,6 +110,7 @@
 
     def sync_changeset(self, rev):
         cset = self.repos.get_changeset(rev)
+        cset = self.cache_mgr.check_changeset(cset)
         srev = self.db_rev(cset.rev)
         old_cset = [None]
 
@@ -251,6 +275,7 @@
                     self.log.info("Trying to sync revision [%s]",
                                   next_youngest)
                     cset = self.repos.get_changeset(next_youngest)
+                    cset = self.cache_mgr.check_changeset(cset)
                     try:
                         # steps 1. and 2.
                         self._insert_changeset(cursor, next_youngest, cset)
